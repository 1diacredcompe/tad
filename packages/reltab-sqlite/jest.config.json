{
  "transform": {
    "^.+\\.tsx?$":  "ts-jest"
  },
  "testRegex": "/test/.*.test.ts",
  "testEnvironment": "node",
  "moduleFileExtensions": [
    "ts",
    "js"
  ]
}
