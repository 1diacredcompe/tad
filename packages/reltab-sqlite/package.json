{
  "name": "reltab-sqlite",
  "version": "0.10.0",
  "description": "sqlite driver for reltab",
  "main": "dist/reltab-sqlite.js",
  "types": "dist/reltab-sqlite.d.ts",
  "scripts": {
    "prepublish": "npm run build",
    "clean": "rimraf dist",
    "build": "tsc",
    "test": "jest --config jest.config.json"
  },
  "keywords": [
    "relational",
    "sql",
    "database",
    "sqlite"
  ],
  "author": "Antony Courtney <antony@antonycourtney.com>",
  "license": "MIT",
  "dependencies": {
    "@types/sqlite3": "^3.1.6",
    "@types/through": "0.0.30",
    "byline": "^5.0.0",
    "csv-sniffer": "^0.1.1",
    "fast-csv": "^4.1.3",
    "gauge": "^3.0.0",
    "loglevel": "^1.6.8",
    "reltab": "^0.10.0",
    "sqlite3": "git+https://github.com/antonycourtney/node-sqlite3.git",
    "through": "^2.3.8",
    "typed-promisify": "^0.4.0"
  },
  "devDependencies": {
    "@types/jest": "^25.2.1",
    "@types/node": "^13.11.0",
    "jest": "^25.2.7",
    "mkdirp": "^1.0.4",
    "prettier": "^2.0.2",
    "ts-jest": "^25.3.1",
    "typescript": "^3.8.3"
  }
}
