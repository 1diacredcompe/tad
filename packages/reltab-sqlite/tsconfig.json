{
    "compilerOptions": {
        "outDir": "./dist",
        "declaration": true,
        "target": "ES5",
        "lib": [
            "es6",
            "es2017",
            "dom"
        ]    
    },
    "include": [
        "./src/*"
    ]
}