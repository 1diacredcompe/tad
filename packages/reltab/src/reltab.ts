export * from "./dialectRegistry";
export * from "./Connection";
export * from "./defs";
export * from "./FilterExp";
export * from "./QueryExp";
export * from "./Schema";
export * from "./TableRep";
export * from "./AggFn";
export * from "./ColumnType";
export * from "./DataSource";
