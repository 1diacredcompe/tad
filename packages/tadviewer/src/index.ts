/*
export * from "./components/AggPanel";
export * from "./components/ColumnList";
export * from "./components/ColumnRow";
export * from "./components/ColumnSelector";
export * from "./components/DisplayOrderPanel";
export * from "./components/FilterEditor";
export * from "./components/FilterEditorRow";
export * from "./components/Footer";
export * from "./components/FormatPanel";
export * from "./components/GridPane";
export * from "./components/IndeterminateCheckbox";
export * from "./components/LoadingModal";
export * from "./components/NumFormatPanel";
export * from "./components/PivotOrderPanel";
export * from "./components/Sidebar";
export * from "./components/SingleColumnSelect";
export * from "./components/SortOrderPanel";
export * from "./components/TextFormatPanel";
*/

export * from "./AppState";
export * from "./components/AppPane";
