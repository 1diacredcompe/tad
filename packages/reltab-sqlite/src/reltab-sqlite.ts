import * as tp from "typed-promisify";
import * as sqlite3 from "sqlite3";
import * as log from "loglevel";
import { TableRep, QueryExp, Schema, tableQuery } from "reltab";
import {
  TableInfoMap,
  TableInfo,
  ValExp,
  Row,
  AggColSpec,
  SubExp,
  ColumnMetaMap,
  ColumnMapInfo,
  ColumnExtendVal,
  Connection,
} from "reltab"; // eslint-disable-line

export * from "./csvimport";

function assertDefined<A>(x: A | undefined | null): A {
  if (x == null) {
    throw new Error("unexpected null value");
  }

  return x;
}

const dbAll = tp.promisify((db, query, cb) => db.all(query, cb));

interface ContextOptions {
  showQueries?: boolean;
}

export class SqliteContext implements Connection {
  db: sqlite3.Database;
  tableMap: TableInfoMap;
  showQueries: boolean;

  constructor(db: any, options: ContextOptions) {
    this.db = db;
    this.tableMap = {};
    this.showQueries = options && options.showQueries;
  }

  registerTable(ti: TableInfo) {
    this.tableMap[ti.tableName] = ti;
  }

  evalQuery(
    query: QueryExp,
    offset: number = -1,
    limit: number = -1
  ): Promise<TableRep> {
    let t0 = process.hrtime();
    const schema = query.getSchema(this.tableMap);
    const sqlQuery = query.toSql(this.tableMap, offset, limit);
    let t1 = process.hrtime(t0);
    const [t1s, t1ns] = t1;

    if (this.showQueries) {
      log.info("time to generate sql: %ds %dms", t1s, t1ns / 1e6);
      log.debug("SqliteContext.evalQuery: evaluating:");
      log.debug(sqlQuery);
    }

    const t2 = process.hrtime();
    const qp = dbAll(this.db, sqlQuery);
    return qp.then((dbRows) => {
      const rows = dbRows as Row[];
      const t3 = process.hrtime(t2);
      const [t3s, t3ns] = t3;
      const t4pre = process.hrtime();
      const ret = new TableRep(schema, rows);
      const t4 = process.hrtime(t4pre);
      const [t4s, t4ns] = t4;

      if (this.showQueries) {
        log.info("time to run query: %ds %dms", t3s, t3ns / 1e6);
        log.info("time to mk table rep: %ds %dms", t4s, t4ns / 1e6);
      }

      return ret;
    });
  }

  rowCount(query: QueryExp): Promise<number> {
    let t0 = process.hrtime();
    const countSql = query.toCountSql(this.tableMap);
    let t1 = process.hrtime(t0);
    const [t1s, t1ns] = t1;

    if (this.showQueries) {
      log.info("time to generate sql: %ds %dms", t1s, t1ns / 1e6);
      log.debug("SqliteContext.evalQuery: evaluating:");
      log.debug(countSql);
    }

    const t2 = process.hrtime();
    const qp = dbAll(this.db, countSql);
    return qp.then((rows) => {
      const t3 = process.hrtime(t2);
      const [t3s, t3ns] = t3;
      log.info("time to run query: %ds %dms", t3s, t3ns / 1e6);
      const ret = Number.parseInt(rows[0].rowCount);
      return ret;
    });
  } // use table_info pragma to construct a TableInfo:

  // Get table info directly from sqlite db
  dbGetTableInfo(tableName: string): Promise<TableInfo> {
    const tiQuery = `PRAGMA table_info(${tableName})`;
    const qp = dbAll(this.db, tiQuery);
    return qp.then((dbRows) => {
      const rows = dbRows as Row[];
      log.debug("getTableInfo: ", rows);

      const extendCMap = (
        cmm: ColumnMetaMap,
        row: any,
        idx: number
      ): ColumnMetaMap => {
        const cnm = row.name;
        const cType = row.type.toLocaleLowerCase();

        if (cType == null) {
          log.error(
            'mkTableInfo: No column type for "' + cnm + '", index: ' + idx
          );
        }

        const cmd = {
          displayName: cnm,
          type: assertDefined(cType),
        };
        cmm[cnm] = cmd;
        return cmm;
      };

      const cmMap = rows.reduce(extendCMap, {});
      const columnIds = rows.map((r) => r.name);
      const schema = new Schema(columnIds as string[], cmMap);
      return {
        tableName,
        schema,
      };
    });
  }

  async getTableInfo(tableName: string): Promise<TableInfo> {
    let ti = this.tableMap[tableName];
    if (!ti) {
      ti = await this.dbGetTableInfo(tableName);
      if (ti) {
        this.tableMap[tableName] = ti;
      }
    }
    return ti;
  }
}

// A wrapper the constructor for sqlite3.Database that returns a Promise.
const open = (filename: string, mode: number): Promise<sqlite3.Database> => {
  return new Promise((resolve, reject) => {
    const db = new sqlite3.Database(filename, mode, (err: Error) => {
      if (err) {
        reject(err);
      }
      resolve(db);
    });
  });
};

const init = async (dbfile, options: Object = {}): Promise<Connection> => {
  log.setLevel(log.levels.DEBUG);
  const db = await open(dbfile, sqlite3.OPEN_READWRITE);
  const ctx = new SqliteContext(db, options);
  return ctx;
};

// get (singleton) connection to sqlite:
let ctxPromise: Promise<Connection> | undefined | null = null;
export const getContext = (
  dbfile: string,
  options: Object = {}
): Promise<Connection> => {
  console.log("getContext: ", dbfile, options);
  if (!ctxPromise) {
    ctxPromise = init(dbfile, options);
  }

  return ctxPromise;
};
