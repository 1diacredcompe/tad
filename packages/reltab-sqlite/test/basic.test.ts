import * as sqlite3 from "sqlite3";
import * as reltab from "reltab";
import * as reltabSqlite from "../src/reltab-sqlite";
import * as tp from "typed-promisify";
import { textSpanContainsPosition } from "typescript";
import { delimiter } from "path";
import * as log from 'loglevel';
import * as util from './testUtils';

const { col, constVal } = reltab;

log.setLevel('debug');

let testCtx: reltabSqlite.SqliteContext;

test("t0 - trivial query generation", () => {
  const q1 = reltab.tableQuery("barttest");

  expect(q1).toMatchInlineSnapshot(`
QueryExp {
  "expType": "QueryExp",
  "operator": "table",
  "tableArgs": Array [],
  "valArgs": Array [
    "barttest",
  ],
}
`);
});

const importCsv = async (db, path) => {
  const md = await reltabSqlite.fastImport(db, path);

  const ti = reltabSqlite.mkTableInfo(md);
  testCtx.registerTable(ti);
}

beforeAll(async (): Promise<reltabSqlite.SqliteContext> => {
  const ctx = await reltabSqlite.getContext(":memory:");

  testCtx = (ctx as reltabSqlite.SqliteContext);

  const db = testCtx.db;

  await importCsv(db, 'test/support/sample.csv');
  await importCsv(db, 'test/support/barttest.csv');

  return testCtx;
})

test("t1 - basic sqlite tableQuery", async () => {
  const q1 = reltab.tableQuery('sample');

  const qres = await testCtx.evalQuery(q1);
  expect(qres).toMatchSnapshot();
});

const bartTableQuery = reltab.tableQuery('barttest');

test("t2 - basic bart table query", async () => {
  const qres = await testCtx.evalQuery(bartTableQuery);

  expect(qres).toMatchSnapshot();
});

test("basic rowcount", async () => {
  const rowCount = await testCtx.rowCount(bartTableQuery);

  expect(rowCount).toBe(23);
});

test('basic project operator', async () => {
  const pcols = ['JobFamily', 'Title', 'Union', 'Name', 'Base', 'TCOE']
  const q2 = bartTableQuery.project(pcols);

  const qres = await testCtx.evalQuery(q2);
  expect(qres.schema.columns).toEqual(pcols);

  expect(qres).toMatchSnapshot();
});

test('basic groupBy', async () => {
  const q3 = bartTableQuery.groupBy(['JobFamily', 'Title'], ['TCOE'])  // note: [ 'TCOE' ] equivalent to [ [ 'sum', 'TCOE' ] ]

  // testCtx.showQueries = true;
  const qres = await testCtx.evalQuery(q3);

  const expCols = ['JobFamily', 'Title', 'TCOE'];
  expect(qres.schema.columns).toEqual(expCols);

  expect(qres.rowData.length).toBe(19);

  const baseRes = await testCtx.evalQuery(bartTableQuery);
  const tcoeSum = util.columnSum(baseRes, 'TCOE');

  const groupSum = util.columnSum(qres, 'TCOE');
  expect(groupSum).toBe(tcoeSum);

  expect(qres).toMatchSnapshot();
});

test('basic filter', async () => {
  const q5 = bartTableQuery.filter(reltab.and().eq(col('JobFamily'), constVal('Executive Management')));

  const res = await testCtx.evalQuery(q5);
  expect(res.rowData.length).toBe(4);

  expect(res).toMatchSnapshot();
});

