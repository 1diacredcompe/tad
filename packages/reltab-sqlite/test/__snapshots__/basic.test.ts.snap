// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic filter 1`] = `
TableRep {
  "rowData": Array [
    Object {
      "Base": 312461,
      "JobFamily": "Executive Management",
      "Name": "Crunican, Grace",
      "TCOE": 399921,
      "Title": "General Manager",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 298700,
      "JobFamily": "Executive Management",
      "Name": "Dugger, Dorothy",
      "TCOE": 419661,
      "Title": "General Manager",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 212647,
      "JobFamily": "Executive Management",
      "Name": "Oversier, Paul",
      "TCOE": 288160,
      "Title": "Assistant General Manager Operations",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 206914,
      "JobFamily": "Executive Management",
      "Name": "DeVaughn, Marcia",
      "TCOE": 281548,
      "Title": "Deputy General Manager",
      "Union": "Non-Represented",
    },
  ],
  "schema": Schema {
    "_sortedColumns": null,
    "columnIndices": Object {
      "Base": 2,
      "JobFamily": 4,
      "Name": 0,
      "TCOE": 3,
      "Title": 1,
      "Union": 5,
    },
    "columnMetadata": Object {
      "Base": Object {
        "displayName": "Base",
        "type": "integer",
      },
      "JobFamily": Object {
        "displayName": "Job Family",
        "type": "text",
      },
      "Name": Object {
        "displayName": "Name",
        "type": "text",
      },
      "TCOE": Object {
        "displayName": "TCOE",
        "type": "integer",
      },
      "Title": Object {
        "displayName": "Title",
        "type": "text",
      },
      "Union": Object {
        "displayName": "Union",
        "type": "text",
      },
    },
    "columns": Array [
      "Name",
      "Title",
      "Base",
      "TCOE",
      "JobFamily",
      "Union",
    ],
  },
}
`;

exports[`basic groupBy 1`] = `
TableRep {
  "rowData": Array [
    Object {
      "JobFamily": "Engineering & Systems Engineering",
      "TCOE": 212535,
      "Title": "Group Manager Systems Capital Prog",
    },
    Object {
      "JobFamily": "Engineering & Systems Engineering",
      "TCOE": 209709,
      "Title": "Group Manager, Capital Projects",
    },
    Object {
      "JobFamily": "Executive Management",
      "TCOE": 288160,
      "Title": "Assistant General Manager Operations",
    },
    Object {
      "JobFamily": "Executive Management",
      "TCOE": 281548,
      "Title": "Deputy General Manager",
    },
    Object {
      "JobFamily": "Executive Management",
      "TCOE": 819582,
      "Title": "General Manager",
    },
    Object {
      "JobFamily": "Finance & Accounting",
      "TCOE": 287139,
      "Title": "Controller-Treasurer",
    },
    Object {
      "JobFamily": "Legal & Paralegal",
      "TCOE": 321543,
      "Title": "General Counsel",
    },
    Object {
      "JobFamily": "Legal & Paralegal",
      "TCOE": 445028,
      "Title": "Senior Attorney",
    },
    Object {
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "TCOE": 113775,
      "Title": "Electrician",
    },
    Object {
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "TCOE": 124962,
      "Title": "Structures Inspector",
    },
    Object {
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "TCOE": 104260,
      "Title": "Transit Vehicle Mechanic",
    },
    Object {
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "TCOE": 123432,
      "Title": "Vehicle Inspector",
    },
    Object {
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "TCOE": 128332,
      "Title": "Wayside Inspector",
    },
    Object {
      "JobFamily": "Police",
      "TCOE": 127660,
      "Title": "Revenue Protection Guard",
    },
    Object {
      "JobFamily": "Public Affairs & Marketing",
      "TCOE": 181955,
      "Title": "Department Manager Gov't & Comm Rel",
    },
    Object {
      "JobFamily": "Safety",
      "TCOE": 211941,
      "Title": "Chief Safety Officer",
    },
    Object {
      "JobFamily": "Transportation Operations",
      "TCOE": 238211,
      "Title": "Chief Transportation Officer",
    },
    Object {
      "JobFamily": "Transportation Operations",
      "TCOE": 352988,
      "Title": "Station Agent",
    },
    Object {
      "JobFamily": "Transportation Operations",
      "TCOE": 118799,
      "Title": "Train Operator",
    },
  ],
  "schema": Schema {
    "_sortedColumns": null,
    "columnIndices": Object {
      "JobFamily": 0,
      "TCOE": 2,
      "Title": 1,
    },
    "columnMetadata": Object {
      "Base": Object {
        "displayName": "Base",
        "type": "integer",
      },
      "JobFamily": Object {
        "displayName": "Job Family",
        "type": "text",
      },
      "Name": Object {
        "displayName": "Name",
        "type": "text",
      },
      "TCOE": Object {
        "displayName": "TCOE",
        "type": "integer",
      },
      "Title": Object {
        "displayName": "Title",
        "type": "text",
      },
      "Union": Object {
        "displayName": "Union",
        "type": "text",
      },
    },
    "columns": Array [
      "JobFamily",
      "Title",
      "TCOE",
    ],
  },
}
`;

exports[`basic project operator 1`] = `
TableRep {
  "rowData": Array [
    Object {
      "Base": 312461,
      "JobFamily": "Executive Management",
      "Name": "Crunican, Grace",
      "TCOE": 399921,
      "Title": "General Manager",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 298700,
      "JobFamily": "Executive Management",
      "Name": "Dugger, Dorothy",
      "TCOE": 419661,
      "Title": "General Manager",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 238269,
      "JobFamily": "Legal & Paralegal",
      "Name": "Burrows, Matthew",
      "TCOE": 321543,
      "Title": "General Counsel",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 212647,
      "JobFamily": "Executive Management",
      "Name": "Oversier, Paul",
      "TCOE": 288160,
      "Title": "Assistant General Manager Operations",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 208744,
      "JobFamily": "Finance & Accounting",
      "Name": "Schroeder, Scott",
      "TCOE": 287139,
      "Title": "Controller-Treasurer",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 206914,
      "JobFamily": "Executive Management",
      "Name": "DeVaughn, Marcia",
      "TCOE": 281548,
      "Title": "Deputy General Manager",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 170383,
      "JobFamily": "Transportation Operations",
      "Name": "Crespo, Rodolfo",
      "TCOE": 238211,
      "Title": "Chief Transportation Officer",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 165470,
      "JobFamily": "Legal & Paralegal",
      "Name": "Lee, Thomas",
      "TCOE": 224345,
      "Title": "Senior Attorney",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 160289,
      "JobFamily": "Legal & Paralegal",
      "Name": "Nuetzel, Victoria",
      "TCOE": 220683,
      "Title": "Senior Attorney",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 157300,
      "JobFamily": "Engineering & Systems Engineering",
      "Name": "Ruffa, Frank",
      "TCOE": 209709,
      "Title": "Group Manager, Capital Projects",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 152434,
      "JobFamily": "Engineering & Systems Engineering",
      "Name": "Dupont, Jean-Luc",
      "TCOE": 212535,
      "Title": "Group Manager Systems Capital Prog",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 152125,
      "JobFamily": "Safety",
      "Name": "Hardy, Leonard",
      "TCOE": 211941,
      "Title": "Chief Safety Officer",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 77446,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Hui, Sam",
      "TCOE": 128332,
      "Title": "Wayside Inspector",
      "Union": "SEIU",
    },
    Object {
      "Base": 77446,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Reinhardt, Michael",
      "TCOE": 124962,
      "Title": "Structures Inspector",
      "Union": "SEIU",
    },
    Object {
      "Base": 77446,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Estrada, Phillip",
      "TCOE": 123432,
      "Title": "Vehicle Inspector",
      "Union": "SEIU",
    },
    Object {
      "Base": 75701,
      "JobFamily": "Police",
      "Name": "Ornellas, Robert",
      "TCOE": 127660,
      "Title": "Revenue Protection Guard",
      "Union": "BPOA",
    },
    Object {
      "Base": 74068,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Smith, Daniel",
      "TCOE": 104260,
      "Title": "Transit Vehicle Mechanic",
      "Union": "SEIU",
    },
    Object {
      "Base": 61184,
      "JobFamily": "Transportation Operations",
      "Name": "Smith, Westley",
      "TCOE": 118799,
      "Title": "Train Operator",
      "Union": "ATU",
    },
    Object {
      "Base": 130035,
      "JobFamily": "Public Affairs & Marketing",
      "Name": "Hamill, Kerry",
      "TCOE": 181955,
      "Title": "Department Manager Gov't & Comm Rel",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 60790,
      "JobFamily": "Transportation Operations",
      "Name": "Serrianni, Cheryl",
      "TCOE": 132030,
      "Title": "Station Agent",
      "Union": "ATU",
    },
    Object {
      "Base": 60781,
      "JobFamily": "Transportation Operations",
      "Name": "Nagel, Kevin",
      "TCOE": 105854,
      "Title": "Station Agent",
      "Union": "ATU",
    },
    Object {
      "Base": 60771,
      "JobFamily": "Transportation Operations",
      "Name": "Riggs, Joshua",
      "TCOE": 115104,
      "Title": "Station Agent",
      "Union": "ATU",
    },
    Object {
      "Base": 59817,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Knight, Joshua",
      "TCOE": 113775,
      "Title": "Electrician",
      "Union": "SEIU",
    },
  ],
  "schema": Schema {
    "_sortedColumns": null,
    "columnIndices": Object {
      "Base": 4,
      "JobFamily": 0,
      "Name": 3,
      "TCOE": 5,
      "Title": 1,
      "Union": 2,
    },
    "columnMetadata": Object {
      "Base": Object {
        "displayName": "Base",
        "type": "integer",
      },
      "JobFamily": Object {
        "displayName": "Job Family",
        "type": "text",
      },
      "Name": Object {
        "displayName": "Name",
        "type": "text",
      },
      "TCOE": Object {
        "displayName": "TCOE",
        "type": "integer",
      },
      "Title": Object {
        "displayName": "Title",
        "type": "text",
      },
      "Union": Object {
        "displayName": "Union",
        "type": "text",
      },
    },
    "columns": Array [
      "JobFamily",
      "Title",
      "Union",
      "Name",
      "Base",
      "TCOE",
    ],
  },
}
`;

exports[`t1 - basic sqlite tableQuery 1`] = `
TableRep {
  "rowData": Array [
    Object {
      "email": "john@doe.com",
      "firstName": "John",
      "lastName": "Doe",
      "phoneNumber": 123456789,
    },
    Object {
      "email": "jane@doe.com",
      "firstName": "Jane",
      "lastName": "Doe",
      "phoneNumber": 9876543210,
    },
    Object {
      "email": "james.bond@mi6.co.uk",
      "firstName": "James",
      "lastName": "Bond",
      "phoneNumber": 612345678,
    },
  ],
  "schema": Schema {
    "_sortedColumns": null,
    "columnIndices": Object {
      "email": 2,
      "firstName": 0,
      "lastName": 1,
      "phoneNumber": 3,
    },
    "columnMetadata": Object {
      "email": Object {
        "displayName": "email",
        "type": "text",
      },
      "firstName": Object {
        "displayName": "firstName",
        "type": "text",
      },
      "lastName": Object {
        "displayName": "lastName",
        "type": "text",
      },
      "phoneNumber": Object {
        "displayName": "phoneNumber",
        "type": "integer",
      },
    },
    "columns": Array [
      "firstName",
      "lastName",
      "email",
      "phoneNumber",
    ],
  },
}
`;

exports[`t2 - basic bart table query 1`] = `
TableRep {
  "rowData": Array [
    Object {
      "Base": 312461,
      "JobFamily": "Executive Management",
      "Name": "Crunican, Grace",
      "TCOE": 399921,
      "Title": "General Manager",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 298700,
      "JobFamily": "Executive Management",
      "Name": "Dugger, Dorothy",
      "TCOE": 419661,
      "Title": "General Manager",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 238269,
      "JobFamily": "Legal & Paralegal",
      "Name": "Burrows, Matthew",
      "TCOE": 321543,
      "Title": "General Counsel",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 212647,
      "JobFamily": "Executive Management",
      "Name": "Oversier, Paul",
      "TCOE": 288160,
      "Title": "Assistant General Manager Operations",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 208744,
      "JobFamily": "Finance & Accounting",
      "Name": "Schroeder, Scott",
      "TCOE": 287139,
      "Title": "Controller-Treasurer",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 206914,
      "JobFamily": "Executive Management",
      "Name": "DeVaughn, Marcia",
      "TCOE": 281548,
      "Title": "Deputy General Manager",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 170383,
      "JobFamily": "Transportation Operations",
      "Name": "Crespo, Rodolfo",
      "TCOE": 238211,
      "Title": "Chief Transportation Officer",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 165470,
      "JobFamily": "Legal & Paralegal",
      "Name": "Lee, Thomas",
      "TCOE": 224345,
      "Title": "Senior Attorney",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 160289,
      "JobFamily": "Legal & Paralegal",
      "Name": "Nuetzel, Victoria",
      "TCOE": 220683,
      "Title": "Senior Attorney",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 157300,
      "JobFamily": "Engineering & Systems Engineering",
      "Name": "Ruffa, Frank",
      "TCOE": 209709,
      "Title": "Group Manager, Capital Projects",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 152434,
      "JobFamily": "Engineering & Systems Engineering",
      "Name": "Dupont, Jean-Luc",
      "TCOE": 212535,
      "Title": "Group Manager Systems Capital Prog",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 152125,
      "JobFamily": "Safety",
      "Name": "Hardy, Leonard",
      "TCOE": 211941,
      "Title": "Chief Safety Officer",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 77446,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Hui, Sam",
      "TCOE": 128332,
      "Title": "Wayside Inspector",
      "Union": "SEIU",
    },
    Object {
      "Base": 77446,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Reinhardt, Michael",
      "TCOE": 124962,
      "Title": "Structures Inspector",
      "Union": "SEIU",
    },
    Object {
      "Base": 77446,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Estrada, Phillip",
      "TCOE": 123432,
      "Title": "Vehicle Inspector",
      "Union": "SEIU",
    },
    Object {
      "Base": 75701,
      "JobFamily": "Police",
      "Name": "Ornellas, Robert",
      "TCOE": 127660,
      "Title": "Revenue Protection Guard",
      "Union": "BPOA",
    },
    Object {
      "Base": 74068,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Smith, Daniel",
      "TCOE": 104260,
      "Title": "Transit Vehicle Mechanic",
      "Union": "SEIU",
    },
    Object {
      "Base": 61184,
      "JobFamily": "Transportation Operations",
      "Name": "Smith, Westley",
      "TCOE": 118799,
      "Title": "Train Operator",
      "Union": "ATU",
    },
    Object {
      "Base": 130035,
      "JobFamily": "Public Affairs & Marketing",
      "Name": "Hamill, Kerry",
      "TCOE": 181955,
      "Title": "Department Manager Gov't & Comm Rel",
      "Union": "Non-Represented",
    },
    Object {
      "Base": 60790,
      "JobFamily": "Transportation Operations",
      "Name": "Serrianni, Cheryl",
      "TCOE": 132030,
      "Title": "Station Agent",
      "Union": "ATU",
    },
    Object {
      "Base": 60781,
      "JobFamily": "Transportation Operations",
      "Name": "Nagel, Kevin",
      "TCOE": 105854,
      "Title": "Station Agent",
      "Union": "ATU",
    },
    Object {
      "Base": 60771,
      "JobFamily": "Transportation Operations",
      "Name": "Riggs, Joshua",
      "TCOE": 115104,
      "Title": "Station Agent",
      "Union": "ATU",
    },
    Object {
      "Base": 59817,
      "JobFamily": "Maintenance, Vehicle & Facilities",
      "Name": "Knight, Joshua",
      "TCOE": 113775,
      "Title": "Electrician",
      "Union": "SEIU",
    },
  ],
  "schema": Schema {
    "_sortedColumns": null,
    "columnIndices": Object {
      "Base": 2,
      "JobFamily": 4,
      "Name": 0,
      "TCOE": 3,
      "Title": 1,
      "Union": 5,
    },
    "columnMetadata": Object {
      "Base": Object {
        "displayName": "Base",
        "type": "integer",
      },
      "JobFamily": Object {
        "displayName": "Job Family",
        "type": "text",
      },
      "Name": Object {
        "displayName": "Name",
        "type": "text",
      },
      "TCOE": Object {
        "displayName": "TCOE",
        "type": "integer",
      },
      "Title": Object {
        "displayName": "Title",
        "type": "text",
      },
      "Union": Object {
        "displayName": "Union",
        "type": "text",
      },
    },
    "columns": Array [
      "Name",
      "Title",
      "Base",
      "TCOE",
      "JobFamily",
      "Union",
    ],
  },
}
`;
